src/externalized/strategic/CreatureLairModel.cc

```diff
@@ -0,0 +1,259 @@
+	for (auto& el : json.GetArray())
+	{
+		JsonObjectReader obj(el);
+
+		auto sectorString = obj.GetString("sector");
+		if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+		{
+			throw new std::runtime_error("Creature attack sector is invalid");
+		}
+
+		CreatureAttackSector sectorAttack;
+		sectorAttack.chance = obj.GetUInt("chance");
+		sectorAttack.insertionCode = insertionCodeFromString(obj.GetString("insertionCode"));
+		sectorAttack.insertionGridNo = obj.getOptionalInt("insertionGridNo");
+		sectorAttack.sectorId = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+		attacks.push_back(sectorAttack);
+	}
+
+	if (attacks.size() == 0)
+	{
+		SLOGE(ST::format("Lair has no town attack sectors"));
+		throw std::runtime_error("");
+	}
+
+	return attacks;
+}
+
+bool CreatureLairModel::isSectorInLair(uint8_t sectorX, uint8_t sectorY, uint8_t sectorZ) const
+{
+	uint8_t sectorId = SECTOR(sectorX, sectorY);
+	for (auto sec : lairSectors)
+	{
+		if (sec.sectorId == sectorId && sec.sectorLevel == sectorZ)
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+const CreatureAttackSector* CreatureLairModel::chooseTownSectorToAttack() const
+{
+	unsigned int totalChance = 0;
+	for (auto sec : attackSectors)
+	{
+		totalChance += sec.chance;
+	}
+
+	int random = Random(totalChance);
+	for (auto& sec : attackSectors)
+	{
+		random -= sec.chance;
+		if (random < 0)
+		{
+			return &sec;
+		}
+	}
+
+	SLOGA("Some coding error with the random selection logic. We should never reach here");
+	return NULL;
+}
+
+const CreatureAttackSector* CreatureLairModel::getTownAttackDetails(uint8_t sectorId) const
+{
+	for (auto& sec : attackSectors)
+	{
+		if (sec.sectorId == sectorId) 
+		{
+			return &sec;
+		}
+	}
+	return NULL;
+}
+
+CreatureLairModel* CreatureLairModel::deserialize(const rapidjson::Value& json)
+{
+	JsonObjectReader obj(json);
+
+	auto entrance = json["entranceSector"].GetArray();
+	auto sectorString = entrance[0].GetString();
+	if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+	{
+		throw new std::runtime_error("Entrance sector is invalid");
+	}
+	uint8_t entranceSector = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+	uint8_t entranceSectorLevel = entrance[1].GetUint();
+
+	auto warpExit = JsonObjectReader(json["warpExit"]);
+	sectorString = warpExit.GetString("sector");
+	if (!IS_VALID_SECTOR_SHORT_STRING(sectorString))
+	{
+		throw new std::runtime_error("Warp exit sector is invalid");
+	}
+	uint8_t warpExitSector = SECTOR_FROM_SECTOR_SHORT_STRING(sectorString);
+	uint16_t warpExitGridNo = warpExit.GetUInt("gridNo");
+
+	return new CreatureLairModel(
+		obj.GetUInt("lairId"),
+		obj.GetUInt("associatedMineId"),
+		entranceSector, entranceSectorLevel,
+		readLairSectors(json["sectors"]),
+		readAttackSectors(json["attackSectors"]),
+		warpExitSector, warpExitGridNo
+	);
+}
+
+void CreatureLairModel::validateData(const std::vector<const CreatureLairModel*> lairs, const std::vector<const UndergroundSectorModel*> ugSectors, const uint8_t numMines)
```

Comment:
You probably wanted to receive const references, but you're receiving const copies.

Reply:
Sometimes I wonder if the compiler is smart enough or allowed to optimize this.